File: service/lib/agama/storage/config_conversions/to_json_conversions/partition.rb
# frozen_string_literal: true

# Copyright (c) [2024] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/storage/config_conversions/to_json_conversions/base"
require "agama/storage/config_conversions/to_json_conversions/with_encryption"
require "agama/storage/config_conversions/to_json_conversions/with_filesystem"
require "agama/storage/config_conversions/to_json_conversions/with_search"
require "agama/storage/config_conversions/to_json_conversions/with_size"
require "agama/storage/configs/partition"

module Agama
  module Storage
    module ConfigConversions
      module ToJSONConversions
        # Partition conversion to JSON hash according to schema.
        class Partition < Base
          include WithSearch
          include WithEncryption
          include WithFilesystem
          include WithSize

          # @see Base
          def self.config_type
            Configs::Partition
          end

        private

          # @see Base#conversions
          def conversions
            return convert_delete if config.delete?

            return convert_delete_if_needed if config.delete_if_needed?

            {
              search:     convert_search,
              alias:      config.alias,
              encryption: convert_encryption,
              filesystem: convert_filesystem,
              size:       convert_size,
              id:         config.id&.to_s
            }
          end

          # @return [Hash]
          def convert_delete
            {
              search: convert_search,
              delete: true
            }
          end

          # @return [Hash]
          def convert_delete_if_needed
            {
              search:         convert_search,
              size:           convert_size,
              deleteIfNeeded: true
            }
          end
        end
      end
    end
  end
end

File: web/src/model/storage/config-model/partition.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import configModel from "~/model/storage/config-model";
import partitionModel from "~/model/storage/partition-model";
import type { ConfigModel, Data, Partitionable } from "~/model/storage/config-model";

function indexByName(device: Partitionable.Device, name: string): number {
  return (device.partitions || []).findIndex((p) => p.name && p.name === name);
}

function indexByPath(device: Partitionable.Device, path: string): number {
  return (device.partitions || []).findIndex((p) => p.mountPath === path);
}

/**
 * Adds a new partition.
 *
 * If a partition already exists in the model (e.g., as effect of using the custom policy), then
 * the partition is replaced.
 * */
function add(
  config: ConfigModel.Config,
  collection: Partitionable.CollectionName,
  index: number,
  data: Data.Partition,
): ConfigModel.Config {
  config = configModel.clone(config);
  const device = configModel.partitionable.find(config, collection, index);

  if (device === undefined) return config;

  // Reset the spacePolicy to the default value if the device goes from unused to used
  if (!configModel.partitionable.isUsed(config, device.name) && device.spacePolicy === "keep")
    device.spacePolicy = null;

  const partition = partitionModel.create(data);
  const partitionIndex = indexByName(device, partition.name);

  if (partitionIndex === -1) device.partitions.push(partition);
  else device.partitions[partitionIndex] = partition;

  return config;
}

function edit(
  config: ConfigModel.Config,
  collection: Partitionable.CollectionName,
  index: number,
  mountPath: string,
  data: Data.Partition,
): ConfigModel.Config {
  config = configModel.clone(config);
  const device = configModel.partitionable.find(config, collection, index);

  if (device === undefined) return config;

  const partitionIndex = indexByPath(device, mountPath);
  if (partitionIndex === -1) return config;

  const oldPartition = device.partitions[partitionIndex];
  const newPartition = { ...oldPartition, ...partitionModel.create(data) };
  device.partitions.splice(partitionIndex, 1, newPartition);

  return config;
}

function remove(
  config: ConfigModel.Config,
  collection: Partitionable.CollectionName,
  index: number,
  mountPath: string,
): ConfigModel.Config {
  config = configModel.clone(config);
  const device = configModel.partitionable.find(config, collection, index);

  if (device === undefined) return config;

  const partitionIndex = indexByPath(device, mountPath);
  device.partitions.splice(partitionIndex, 1);

  // Do not delete anything if the device is not really used
  if (!configModel.partitionable.isUsed(config, device.name)) {
    device.spacePolicy = "keep";
  }

  return config;
}

export default { add, edit, remove };
